confmap/resolver.go

```diff
@@ -187,3 +188,59 @@ func (mr *Resolver) closeIfNeeded(ctx context.Context) error {
 	}
 	return err
 }
+
+func (mr *Resolver) expandValue(ctx context.Context, value interface{}) (interface{}, error) {
+	switch v := value.(type) {
+	case string:
+		// If it doesn't have the format "${scheme:opaque}" no need to expand.
+		if !strings.HasPrefix(v, "${") || !strings.HasSuffix(v, "}") {
+			return value, nil
+		}
+		location := v[2 : len(v)-1]
+		// For backwards compatibility:
+		// - empty scheme means "env".
+		ret, err := mr.retrieveValue(ctx, location, "env")
+		if err != nil {
+			return nil, err
+		}
+		mr.closers = append(mr.closers, ret.Close)
+		return ret.AsRaw()
```

Comment:
To confirm: we don't want to apply `expandValue` recursively to the returned result here?

Reply:
How many times would we do this? Until no changes (can get into infinite loop, maybe try for 10 times if still resolving then return error?)?

I thought it is not useful, but I can see now point in having it (after you asked for it). Let's discuss the behavior.

Reply:
I can imagine 2 use cases:
- If you have a local config file where env variables are used (quite common, e.g. in Splunk distro) you may want to specify that the value must be retrieved from a remote source instead of providing it directly.
- The opposite: a config fetched from a remote location, but which refers to a local env variables (e.g. refer to $PATH or $TEMP).

>How many times would we do this? Until no changes (can get into infinite loop, maybe try for 10 times if still resolving then return error?)?

Yeah, probably repeat until no more changes with some limit to avoid infinite looping. Any reasonable use case I can imagine should be 2 rounds, so limiting to something like 5 or 19 attempts should be fine.

Reply:
I would suggest not introducing recursive expanding on this PR. We can forbid usage of `$`, `{` and `}` in the URI on this PR and deal with this problem later

Reply:
There are two problems, and probably we should address both somehow:
1. Embedding `${http://my.example.com?os=${env:OS_ENV}}`
2. Recursive `${file:/path/to/my/file}` which returns a section of the config as a `map[string]interface{}` for example like `{"key": ${env:VALUE}}`. Then should we expand `${env:VALUE}`?

Reply:
@jeromeinsf I think everyone agrees on leaving the implementation of (1) for later and restricting the allowed characters now. For (2), I believe @bogdandrutu wants to add support now with some recursion limit